File: Lab4-middleware-hw/A2-subscription/Server/Services/GrpcServerRunner.cs
namespace SmartHome.Server
{
    public class GrpcServerRunner : IGrcpServerRunner
    {
        private readonly int _port;
        private readonly IDeviceManager _deviceManager;
        private Server _server;

        public GrpcServerRunner(int port, IDeviceManager deviceManager)
        {
            _port = port;
            _deviceManager = deviceManager;
        }

        public Task StartAsync(CancellationToken cancellationToken)
        {
            _server = new Server
            {
                Services = { SmartHomeService.BindService(new SmartHomeServiceImpl(_deviceManager)) },
                Ports = { new ServerPort("0.0.0.0", _port, ServerCredentials.Insecure) }
            };
            _server.Start();
            Console.WriteLine($"gRPC server listening on port {_port}");
            return Task.CompletedTask;
        }

        public Task StopAsync(CancellationToken cancellationToken)
        {
            _server.ShutdownAsync().Wait(cancellationToken);
            Console.WriteLine("gRPC server stopped");
            return Task.CompletedTask;
        }
    }
    }

File: Lab4-middleware-hw/A2-subscription/ServerJava/src/main/java/com/smarthome/server/GrpcServerRunner.java
package com.smarthome.server;

import io.grpc.Server;
import io.grpc.ServerBuilder;
import java.io.IOException;

public class GrpcServerRunner implements IGrpcServerRunner {
    private final int port;
    private final IDeviceManager deviceManager;
    private Server server;

    public GrpcServerRunner(int port, IDeviceManager deviceManager) {
        this.port = port;
        this.deviceManager = deviceManager;
    }

    @Override
    public void start() throws IOException {
        server = ServerBuilder.forPort(port)
                .addService(new SmarthomeServiceImpl(deviceManager))
                .build()
                .start();
        System.out.println("gRPC server listening on port " + port);
    }

    @Override
    public void stop() throws InterruptedException {
        if (server != null) {
            server.shutdown().awaitTermination();
            System.out.println("gRPC server stopped");
        }
    }
}

